[SR-6706] Swift should complain about weak references to classes that don't support them

ChristopherRogers opened the pull request:
<!-- What's in this pull request? -->
This pull requests adds diagnostics similar to Clang when attempting to create weak or unowned references of types that do not support them (NSFont, NSWindow, etc.). This is indicated with the `objc_arc_weak_reference_unavailable` Clang attribute.

 I thought about limiting this to only run on macOS since in practice this only affects macOS targets. (The only classes affected are from the macOS SDK). I've left it for completion's sake since someone could add the attribute to their own Objective-C class if they so choose..

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-6706](https://bugs.swift.org/browse/SR-6706).

ChristopherRogers committed:
Swift should complain about weak references to classes that don't support them

Resolves SR-6706

ChristopherRogers committed:
Typo/grammar fixes

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2193,6 +2195,24 @@ void TypeChecker::checkReferenceOwnershipAttr(VarDecl *var,
+  ClassDecl *underlyingClass = underlyingType->getClassOrBoundGenericClass();
+  while (underlyingClass != nullptr) {
+    if (auto clangDecl = underlyingClass->getClangDecl()) {
+      if (auto underlyingObjCDecl = cast<clang::ObjCInterfaceDecl>(clangDecl)) {
```
Can this cast ever fail, eg if you have a CF type or DispatchQueue or something?

ChristopherRogers replied:
Ah, I didn't think of that. Looks like it would fail then. My original intention was to use `dyn_cast` just in case some new language bridge was added (C++? lol). Thanks.

About CF types & Dispatch objects, `isArcWeakrefUnavailable()` is only defined on `clang::ObjCInterfaceDecl` so I assume the attribute it tests for cannot be added to anything else--or at least Clang doesn't support it at the moment.

jrose-apple commented:
@ChristopherRogers talked to me about this at try! Swift San Jose today. Sorry for losing the PR for almost a year! To avoid having Sema reason about Clang declarations, I suggested adding a bit to ClassDecl to track this, which ClangImporter can set and Sema can check.

ChristopherRogers committed:
Remove duplicate include

ChristopherRogers commented:
Some notes:
- At first I thought maybe I should just set the bit correctly in the constructor but that triggered an assertion so I assume that was the wrong way to go about it.
- I usually like avoiding negatives for boolean names to avoid potential double negatives but in this case it seems that there's a convention of having the default always be zero so I prioritized that (and the wording wasn't that awful).
- I changed the diagnostic message (which was based on the Clang message) to use the word "reference" instead of "variable".

I think I've addressed everything. Let me know if you have any other concerns or nitpicks. Thank you for your help yesterday!

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -3860,6 +3864,17 @@ class ClassDecl final : public NominalTypeDecl {
+  /// references.
+  /// 
+  /// Note that this is true if this class or any of its ancestor classes
+  /// are marked incompatible.
```
Nice, this does make sense to package up in the accessor.

jrose-apple reviewed (approved):
Thanks, @ChristopherRogers! I still think it'd be a good idea to check an Objective-C subclass of `NoWeakClass` as well, but this looks good regardless.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
And while I don't see how it could cause a problem, let's make sure…

@swift-ci Please test source compatibility

ChristopherRogers committed:
Fix broken test on Linux

ChristopherRogers commented:
I had forgotten to require objc_interop for the test file so it was failing on Linux. The macOS compatibility suite had some error with a ClockKit module not being found, but the previous build job had the same error as well.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
I think this is still going to fail but I want to at least see the failures.

@swift-ci Please test source compatibility

jrose-apple commented:
Source compat failures match those on master. Thanks, @ChristopherRogers!
